### PR TITLE
Add ens/zil lookup api

### LINKED ISSUE #388: Implement ENS name resolving

### LINKED ISSUE #452: Implement ZIL name resolving
https://docs.unstoppabledomains.com/#tag/domain_name_resolution/paths/~1{domain}/get

### PR DESCRIPTION
**Pull request by hewigovens**
Fixes #388, Fixes #452 
## ENS 

```bash
curl "http://127.0.0.1:8420/ns/lookup?name=ohmyname.eth&coin=61"
```

```json
{
  "result": "0x12B896601fC07ABF8Edc2f8d94Fd45F042BF7FFD",
  "coin": 61
}
```

## ZIL

```bash
curl "http://127.0.0.1:8420/ns/lookup?name=brad.zil&coin=60"
```

```json
{
  "result":"1EVt92qQnaLDcmVFtHivRJaunG2mf2C3mB",
  "coin": 0
}
```

```json
{
  "result": "0x45b31e01AA6f42F0549aD482BE81635ED3149abb",
  "coin": 60
}
```

### CODE REVIEW

**Comment by hewigovens** [comment]
@Pantani integrate tests failed, maybe we can setup two pipelines or tasks? UT, Integration test

File: api/naming_service.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+		Type  string `json:"type"`
+	} `json:"meta"`
+}
+
+func MakeLookupRoute(router gin.IRouter) {
+	ns := router.Group("/ns")
+	ns.GET("/lookup", handleLookup)
+}
+
+func handleLookup(c *gin.Context) {
```
[the remark this answers is not in the document]
Comment (hewigovens) [reply api/naming_service.go]: added

File: api/naming_service.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	if name == "" {
+		RenderError(c, http.StatusBadRequest, "name query is missing")
+		return
+	}
+	coin, err := strconv.ParseUint(coinQuery, 10, 64)
+	if err != nil {
+		RenderError(c, http.StatusBadRequest, "coin query is invalid")
+		return
+	}
+	if strings.Contains(name, ".eth") {
```
Comment (vikmeup) [inline api/naming_service.go]: should we check for suffix? I think if you type test.eth.zil it will trigger as true.

File: api/naming_service.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+// @Failure 500 {object} api.ApiError
+// @Router /ns/lookup [get]
+func MakeLookupRoute(router gin.IRouter) {
+	ns := router.Group("/ns")
+	ns.GET("/lookup", handleLookup)
+}
+
+func handleLookup(c *gin.Context) {
+	name := c.Query("name")
+	coinQuery := c.DefaultQuery("coin", strconv.Itoa(CoinType.ETH))
```
Comment (vikmeup) [inline api/naming_service.go]: why default to ETH?

File: api/naming_service.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	CoinType "github.com/trustwallet/blockatlas/coin"
+	"github.com/trustwallet/blockatlas/pkg/blockatlas"
+
+	"github.com/ethereum/go-ethereum/ethclient"
+	ens "github.com/wealdtech/go-ens/v3"
+)
+
+type ZNSResponse struct {
+	Addresses map[string]string
+	Meta      struct {
```
Comment (vikmeup) [inline api/naming_service.go]: can we avoid nested structs? 

did you consider splitting logic into separate files? for Zil, ETH and FIO in the future?

File: api/naming_service.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+// @Success 200 {object} blockatlas.Resolved
+// @Failure 500 {object} api.ApiError
+// @Router /ns/lookup [get]
+func MakeLookupRoute(router gin.IRouter) {
+	ns := router.Group("/ns")
+	ns.GET("/lookup", handleLookup)
+}
+
+func handleLookup(c *gin.Context) {
+	name := c.Query("name")
```
Comment (vikmeup) [inline api/naming_service.go]: might worth lowercasing the name as well

File: api/naming_service.go (hunk 4 above)
Comment (hewigovens) [reply api/naming_service.go]: good suggestion, will try to split it into different platforms

Comment (hewigovens) [reply api/naming_service.go]: done

**Comment by hewigovens** [comment]
Will add Bitcoin/BNB later

**Comment by kolya182** [comment]
@hewigovens For Ethereum response address in lowercase and for zil in checksum is that correct ?

**Comment by vikmeup** [comment]
@hewigovens did you consider extending `lookup?name=ohmyname.eth&coin=61` in the future to support `lookup?name=ohmyname.eth` to show list of addresses? to easier read information without breaking an API

File: platform/ethereum/ens.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+		// return empty result for errors like: no unregistered
+		return &result, nil
+	}
+
+	result.Result = mapAddress(coin, address)
+	return &result, nil
+}
+
+func mapAddress(coin uint64, bytes []byte) string {
+	// FIXME: convert bytes to string according to coin
```
Comment (Pantani) [inline platform/ethereum/ens.go]: Do we already have an issue with that? Use TODO, because the bot automatically create an issue for us

Comment (hewigovens) [reply platform/ethereum/ens.go]: no issue yet, I was planning to merge this first and create another one, looks like I can continue working on this branch

**Comment by hewigovens** [comment]
@vikmeup batch query is a bit tricky, it can be different coins, so I would rather have a separate post api

File: api/naming_service.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+var TLDMapping = map[string]uint64{}
+
+// @Summary Lookup .eth / .zil addresses
+// @ID lookup
+// @Description Lookup ENS/ZNS to find registered addresses
+// @Produce json
+// @Tags ns
+// @Param name query string empty "string name"
+// @Param coin query string 60 "string coin"
+// @Success 200 {object} blockatlas.Resolved
```
Comment (vikmeup) [inline api/naming_service.go]: `blockatlas.Resolve` maybe like this?

**Comment by hewigovens** [comment]
or we can refactor to support batch from day one?

**Comment by vikmeup** [comment]
no need, good as is

**Comment by hewigovens** [comment]
@kolya182 will integrate https://github.com/hewigovens/go-coincodec, so it will be checksumed

File: platform/ethereum/ens.go (hunk 6 above)
Comment (hewigovens) [reply platform/ethereum/ens.go]: done

File: api/naming_service.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/gin-gonic/gin"
+	CoinType "github.com/trustwallet/blockatlas/coin"
+	"github.com/trustwallet/blockatlas/platform"
+)
+
+var TLDMapping = map[string]uint64{}
+
+// @Summary Lookup .eth / .zil addresses
+// @ID lookup
+// @Description Lookup ENS/ZNS to find registered addresses
```
Comment (vikmeup) [inline api/naming_service.go]: @Pantani suggestion to rename `coin` to `CoinType`

File: platform/zilliqa/zns.go [hunk 9]
```diff
@@ -0,0 +1,26 @@
+package zilliqa
```
Comment (vikmeup) [inline platform/zilliqa/zns.go]: rename `zns.go` to `ns.go`

File: platform/ethereum/ens.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
+package ethereum
```
Comment (vikmeup) [inline platform/ethereum/ens.go]: rename `ens.go` to `ns.go`

File: pkg/integration/testdata/exclude.json [hunk 11]
```diff
@@ -1,5 +1,6 @@
 [
   "/swagger/*any",
+  "/ns/lookup",
```
Comment (vikmeup) [inline pkg/integration/testdata/exclude.json]: any reason for exclude? can we add integration testing for ns?

Comment (hewigovens) [reply pkg/integration/testdata/exclude.json]: currently integration tests only under specific platform, @Pantani may have more insights

File: platform/ethereum/ens.go (hunk 10 above)
Comment (hewigovens) [reply platform/ethereum/ens.go]: ens is fine, we probably will have `.crypto` resolving, which is not ens powered

File: platform/zilliqa/zns.go (hunk 9 above)
Comment (hewigovens) [reply platform/zilliqa/zns.go]: same as `ens.go`

**Pull request merged**